### PR TITLE
Allow duplicate metric name when meter is different

### LINKED ISSUE #2627: Instrument name collisions for different Meters
Opening this as a tracking issue for the bug where `Instrument`s that have the same name are dropped, even if they come from separate `Meter`s. From the [API spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument): 

> Different Meters MUST be treated as separate namespaces. The names of the Instruments under one Meter SHOULD NOT interfere with Instruments under another Meter.

This issue does **only** cover the fact that duplicate instrument elimination in the .NET SDK implementation does not take into account the `Meter` name, but only the `Instrument` name. It does **not** cover downstream implications (such as how to deal with Prometheus export in such a case, which has extensively been discussed and resolved (?) [elsewhere](https://github.com/open-telemetry/opentelemetry-specification/pull/2035#issuecomment-965540896)).

While this should occur only rarely, I think its important that this is addressed. As more and more libraries adopt OpenTelemetry, such situations are likely to appear more often. OTLP is able to transport these data streams, so it seems counter-intuitive to drop them.

## Additional Context:

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument

https://github.com/open-telemetry/opentelemetry-specification/issues/2064
[Code where instruments are dropped when using Views](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/MeterProviderSdk.cs#L169-L175)
[Code where instruments are dropped without Views](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/MeterProviderSdk.cs#L252)

### PR DESCRIPTION
**Pull request by alanwest**
Fixes #2627.

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 ## Unreleased
 
+* Metrics with the same name but from different meters is allowed.
```
Comment (reyang) [inline src/OpenTelemetry/CHANGELOG.md:5]: nit:

```suggestion
* Metrics with the same name but from different meters are allowed.
```

**Review by reyang — APPROVED** [review]
LGTM.

File: test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs [hunk 2]
```diff
@@ -102,9 +102,11 @@ public void ObserverCallbackExceptionTest()
...
 
         [Theory]
-        [InlineData(AggregationTemporality.Cumulative)]
-        [InlineData(AggregationTemporality.Delta)]
-        public void StreamNamesDuplicatesAreNotAllowedTest(AggregationTemporality temporality)
+        [InlineData(AggregationTemporality.Cumulative, true)]
+        [InlineData(AggregationTemporality.Cumulative, false)]
+        [InlineData(AggregationTemporality.Delta, true)]
+        [InlineData(AggregationTemporality.Delta, false)]
+        public void StreamNamesDuplicatesAreNotAllowedTest(AggregationTemporality temporality, bool hasView)
```
Comment (utpilla) [inline test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs:109]: Rename this to `StreamNameDuplicatesAreAllowedWithDifferentMetersTest`?

We should add another unit test to confirm that we cannot add duplicate metric stream names when the instruments created from the same meter.

**Review by utpilla — APPROVED** [review]
Left the suggestions for unit tests.

File: test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs (hunk 2 above)
Comment (alanwest) [reply test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs:109]: This unit test currently tests both

1) that duplicate names are not allowed in when the instruments are created from the same meter, and
2) that duplicate names are allowed when instruments are from different meters

Are you suggesting breaking this into two tests for clarity? If so, I'm 👍 on that idea.

Comment (utpilla) [reply test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs:109]: > We should also add another unit test to confirm that we cannot add duplicate metric stream names when the instruments created from the same meter.

Ohh I just realized that this was already covered before this PR. I guess we could have two different unit tests so that their names match what they are testing.

Comment (alanwest) [reply test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs]: Split em up. ✂️

**Review by cijothomas — APPROVED** [review]

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs [hunk 3]
```diff
@@ -153,12 +153,14 @@ internal sealed class MeterProviderSdk : MeterProvider
                         for (int i = 0; i < maxCountMetricsToBeCreated; i++)
                         {
                             var metricStreamConfig = metricStreamConfigs[i];
-                            var metricStreamName = metricStreamConfig?.Name ?? instrument.Name;
+                            var meterName = instrument.Meter.Name;
+                            var metricName = metricStreamConfig?.Name ?? instrument.Name;
+                            var metricStreamName = $"{meterName}.{metricName}";
 
-                            if (!MeterProviderBuilderSdk.IsValidInstrumentName(metricStreamName))
+                            if (!MeterProviderBuilderSdk.IsValidInstrumentName(metricName))
```
Comment (pirgeo) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:160]: Shouldn't this be `metricStreamName` still? As it is now, it again uses the instrument name, right?

Comment (utpilla) [reply src/OpenTelemetry/Metrics/MeterProviderSdk.cs:160]: We do want to use the instrument name as the validation check is only for the instrument name. `metricStreamName` is now used to ensure the uniqueness of the combination of meter name and instrument name.

Comment (pirgeo) [reply src/OpenTelemetry/Metrics/MeterProviderSdk.cs:160]: :+1:, thanks for tackling this so quickly!
